includes/validation/class-amp-site-validation.php

```diff
@@ -0,0 +1,279 @@
+
+	/**
+	 * The total number of unaccepted validation errors.
+	 *
+	 * If an error has been accepted in the /wp-admin validation UI,
+	 * it won't count toward this.
+	 *
+	 * @var int
+	 */
+	public static $unaccepted_errors = 0;
+
+	/**
+	 * The number of URLs crawled, regardless of whether they have validation errors.
+	 *
+	 * @var int
+	 */
+	public static $number_crawled = 0;
+
+	/**
+	 * Inits the class.
+	 */
+	public static function init() {
+		if ( defined( 'WP_CLI' ) && WP_CLI ) {
+			WP_CLI::add_command( 'amp', array( __CLASS__, 'crawl_site' ) );
+		}
+	}
+
+	/**
+	 * Crawls the entire site to validate it, and gets the results.
+	 *
+	 * @param array $args The arguments for the command.
+	 */
+	public static function crawl_site( $args ) {
+		if ( ! isset( $args[0] ) || ! self::WP_CLI_ARGUMENT === $args[0] ) {
+			return;
+		}
+
+		$number_urls_to_crawl = self::count_posts_and_terms();
+		WP_CLI::log( sprintf( __( 'Crawling the entire site for AMP validity.', 'amp' ) ) );
+
+		self::$wp_cli_progress = WP_CLI\Utils\make_progress_bar(
+			/* translators: %d is the number of URLs */
+			sprintf( __( 'Validating %d URLs...', 'amp' ), $number_urls_to_crawl ),
+			$number_urls_to_crawl
+		);
+		self::validate_entire_site_urls();
+		self::$wp_cli_progress->finish();
+
+		$url_more_details = add_query_arg(
+			'post_type',
+			\AMP_Invalid_URL_Post_Type::POST_TYPE_SLUG,
+			admin_url( 'edit.php' )
+		);
+
+		WP_CLI::success(
+			sprintf(
+				/* translators: $1%d is the number of URls crawled, $2%d is the number of validation issues, $3%d is the number of unaccepted issues, $4%s is the link for more details */
+				__( "Of the %1\$d URLs crawled, %2\$d have AMP validation issue(s), and %3\$d have unaccepted issue(s).\nFor more details, please see: \n%4\$s", 'amp' ),
+				self::$number_crawled,
+				self::$total_errors,
+				self::$unaccepted_errors,
+				$url_more_details
+			)
+		);
+	}
+
+	/**
+	 * Finds the total number of posts and terms on the site.
+	 *
+	 * The access level is private,
+	 * as this passes 'posts_per_page' => -1 to WP_Query().
+	 * This is only intended for use with WP-CLI,
+	 * and it's a performance risk otherwise.
+	 *
+	 * @return int The number of posts and terms.
+	 */
+	private static function count_posts_and_terms() {
+		$total_count       = 0;
+		$public_post_types = get_post_types( array( 'public' => true ), 'names' );
+		$term_query        = new WP_Term_Query( array(
+			'taxonomy' => get_taxonomies( array( 'public' => true ) ),
+			'fields'   => 'ids',
+		) );
+		$total_count      += count( $term_query->terms );
+
+		/**
+		 * Because of the posts_per_page => -1 value, this is only suited for use in WP-CLI.
+		 * This is necessary to pass as an argument to WP_CLI\Utils\make_progress_bar(),
+		 * in order to show the progress of every post.
+		 */
+		$post_query   = new WP_Query( array(
+			'post_type'      => $public_post_types,
+			'fields'         => 'ids',
+			'posts_per_page' => -1,
```

Comment:
`'posts_per_page' => -1,` is a little troubling, as it could crash this when a site has tens or hundreds of thousands of posts.

Reply:
Indeed. It should do batches of posts at 100 per query, and keep querying while incrementing the page until there are none left. I suggest sorting posts by ID in ascending order, as that will ensure newly published posts don't result in accidental skipping of posts.

Reply:
Thanks, good idea.